File: Libs/TauMPC/src/mpc/finite_field_math/Zp.java
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */

package mpc.finite_field_math;

import java.util.List;
import mpc.circuit.Gate.Operation;


public class Zp {
    public final int prime;
    private int num;

    public Zp(int prime, int num) {
        this.prime = prime;
        this.num = modulo(num , prime);
    }

    public Zp(Zp toCopy) {
        this.prime = toCopy.prime;
        this.num = modulo(toCopy.num , this.prime);
    }
    
    public Zp add(Zp operand2){
        num = modulo(num + operand2.getValue()) ;
        return this;        
    }

  public Zp constAdd(Zp operand2){
        Zp temp = new Zp(this);
        temp.num = modulo(num + operand2.getValue());
        return temp;
    }


    public Zp addListContent(List<Zp> zpList){
        for (Zp zp : zpList){
            this.add(zp);
        }
        return this;
    }

    public Zp sub(Zp operand2){
        num = modulo(num - operand2.getValue()) ;
        return this;
    }

    public Zp constSub(Zp operand2){
        Zp temp = new Zp(this);
        temp.num = modulo(num - operand2.getValue());
        return temp;
    }


    public Zp divide(Zp operand2){
        if (operand2.num == 0)
            throw new IllegalArgumentException( "Cannot divide in zero" );
        num = modulo(num * getFieldInverse(operand2.getValue()) ) ;
        return this;
    }

   public Zp constDvide(Zp operand2){
       if (operand2.num == 0)
            throw new IllegalArgumentException( "Cannot divide in zero" );
        Zp temp = new Zp(this);
        temp.num = modulo(num * getFieldInverse(operand2.getValue()) ) ;
        return temp;
    }


   public  int  getFieldInverse(int fieldNum){
       int temp;
       fieldNum = modulo(fieldNum);
       for (int j=1; j < prime; j++)
       {
               temp = modulo(fieldNum  *  j) ;
               if  (temp == 1)
                       return j;
       }
        return  0;
    }


    public Zp mul(Zp operand2){
        num = modulo(num * operand2.getValue());
        return this;        
    }

    public Zp constMul(Zp operand2){
        Zp temp = new Zp(this);
        temp.num = modulo(num * operand2.getValue());
        return temp;
    }
    
    public Zp Calculate(Zp operand2, Operation operation){
        switch(operation){
            case ADD:
                return add(operand2);
            case MUL:
                return mul(operand2);
           case SUB:
               return sub(operand2);
            case DIV:
                return divide(operand2);
            default:
                assert false;
                throw new RuntimeException("Unknown operation type " + operation);
        }        
    }
        
    public int getValue() {
        return num;
    }

    public void setValue(int newNum) {
        num = modulo(newNum);
    }


    public  int  modulo(int i){
            return  modulo(i , prime);
    }

   public static int  modulo(int i, int prime){
        int temp = i;
        temp = temp % prime;
            if (temp < 0)
                temp += prime;

        return temp;
   }


  public static boolean  isPrime(int num){
        int temp;
        for (int i=2 ; i <= (int) Math.sqrt(num); i++)
        {
                temp = num / i;

                if ( num == (temp * i))
                {
                    return false;
                }
        }
        return true;
  }


 public static int[] getFieldInverseArr(int prime){
        int temp;
        int[] invArr = new int[prime];

        for (int i=0; i< prime; i++)
            for (int j=0; j < prime; j++)
            {
                   temp = modulo(i * j, prime) ;
                   if  (temp == 1)
                   {
                       invArr[i] = j;
                       break;
                   }           
          }
        return invArr;
    }

     public static int  getFieldMinimumPrimitive(int prime){
         int  w_i;
         boolean cond;
         boolean[] fieldElements = new boolean[prime];

        for (int w = 2; w < prime; w++)
        {
                w_i = 1;
                cond = true;
                for (int i = 1; i < prime; i++)
                {
                        fieldElements[i] = false;
                }

                for (int i = 1; i < prime; i++)
                {
                        w_i = modulo(w * w_i, prime);
                        if  (fieldElements[w_i])
                        {
                            cond = false;
                            break;
                        }
                        fieldElements[w_i]  = true;
                }

                if (cond)
                {
                    return w;
                }
        }
        
        throw new IllegalArgumentException("Cannot find field primitive for a field from non prime number.");
     }


     public static int calculatePower(int base , int exp, int prime) {
         int w_i = 1;
         for (int i = 0; i < exp; i++)
         {
                w_i = modulo(base * w_i, prime);
         }
         return w_i;
     }


    public static Zp evalutePolynomialAtPoint(List<Zp> polynomial, Zp point){
         int evaluation = 0;
        for (int i =0; i < polynomial.size();i++){
            evaluation += polynomial.get(i).getValue() * calculatePower(point.getValue(), i, point.prime);
        }
         return new Zp(point.prime, evaluation);
    }
    

    @Override
    public String toString() {
        return String.valueOf(num);
    }

    @Override
    public boolean equals(Object obj) {
        if (!(obj instanceof Zp)){
            return false;
        }
        Zp compare = (Zp)obj;
        return getValue() == compare.getValue() && prime == compare.prime;
    }

    @Override
    public int hashCode() {
        return new Integer(getValue()).hashCode();
    }

    public static void main(String[] args) {
        int prime = 233;

        Zp num1 = new Zp (prime, 4);
        Zp num2 = new Zp (prime, 6);
        int primitive = getFieldMinimumPrimitive(prime);
        System.out.println(primitive);

        int w_i = 1;
        for (int i = 1; i < prime; i++)
        {
                System.out.print( w_i + ", ");
                w_i = modulo(primitive * w_i, prime);
        }
        System.out.println( );


    }


}

File: Libs/FiniteField/Zp.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Numerics;

namespace MpcLib.Common.FiniteField
{
	/// <summary>
	/// Represents a finite field.
	/// </summary>
	public class Zp : ISizable, IEquatable<Zp>
	{
		private int num;
        private BigInteger bigNum;

		public readonly int Prime;

		public int Value
		{
			get
			{
				return num;
			}

			set
			{
				num = Modulo(value);
			}
		}

		public Zp AdditiveInverse
		{
			get
			{
				return new Zp(Prime, -1 * this.num);
			}
		}

		/// <summary>
		/// Multiplicative inverse of the number modulo Prime.
		/// </summary>
		public Zp MultipInverse
		{
			get
			{
				return new Zp(Prime, NumTheoryUtils.MultiplicativeInverse(this.num, Prime));
			}
		}

		public Zp(int prime, int num)
		{
			this.Prime = prime;
			this.num = Modulo(num, prime);
		}

		public Zp(int prime)
		{
			this.Prime = prime;
		}

		public Zp(Zp toCopy)
		{
			this.Prime = toCopy.Prime;
			this.num = Modulo(toCopy.num, this.Prime);
		}

		public static Zp operator +(Zp zp1, Zp zp2)
		{
			Debug.Assert(zp1.Prime == zp2.Prime, "Finite fields do not match!");
			return new Zp(zp1.Prime, zp1.num + zp2.num);
		}

		public static Zp operator -(Zp zp1, Zp zp2)
		{
			Debug.Assert(zp1.Prime == zp2.Prime, "Finite fields do not match!");
			return new Zp(zp1.Prime, zp1.num - zp2.num);
		}

		public static Zp operator *(Zp zp1, Zp zp2)
		{
			Debug.Assert(zp1.Prime == zp2.Prime, "Finite fields do not match!");
			return new Zp(zp1.Prime, zp1.num * zp2.num);
		}

		public static Zp operator /(Zp zp1, Zp zp2)
		{
			Debug.Assert(zp1.Prime == zp2.Prime, "Finite fields do not match!");
			return zp1 * zp2.MultipInverse;
		}

		public static Zp operator +(Zp zp1, int i)
		{
			return new Zp(zp1.Prime, zp1.num + i);
		}

		public static Zp operator ++(Zp zp1)
		{
			zp1.num = Modulo(zp1.num + 1, zp1.Prime);
			return zp1;
		}

		public static bool operator <(Zp zp1, int i)
		{
			return zp1.num < i;
		}

		public static bool operator >(Zp zp1, int i)
		{
			return zp1.num > i;
		}

		public static bool operator ==(Zp zp1, Zp zp2)
		{
			if (Object.ReferenceEquals(zp1, zp2))
				return true;

			if (((object)zp1 == null) || ((object)zp2 == null))
				return false;

			return zp1.num == zp2.num;
		}

		public static bool operator !=(Zp zp1, Zp zp2)
		{
			return !(zp1 == zp2);
		}

		public Zp Add(Zp operand)
		{
			num = Modulo(num + operand.Value);
			return this;
		}

		public Zp ConstAdd(Zp operand)
		{
			var temp = new Zp(this);
			temp.num = Modulo(num + operand.Value);
			return temp;
		}

		public Zp AddListContent(IList<Zp> zpList)
		{
			foreach (Zp zp in zpList)
				Add(zp);

			return this;
		}

		public Zp Sub(Zp operand)
		{
			num = Modulo(num - operand.Value);
			return this;
		}

		public Zp ConstSub(Zp operand)
		{
			var temp = new Zp(this);
			temp.num = Modulo(num - operand.Value);
			return temp;
		}

		public Zp Div(Zp operand)
		{
			if (operand.num == 0)
				throw new ArgumentException("Cannot divide by zero!");

			return this * operand.MultipInverse;
		}

		public Zp ConstDvide(Zp operand)
		{
			if (operand.num == 0)
				throw new System.ArgumentException("Cannot divide by zero!");

			return this * operand.MultipInverse;
		}

		public Zp Mul(Zp operand)
		{
			num = Modulo(num * operand.Value);
			return this;
		}

		public Zp ConstMul(Zp operand)
		{
			var temp = new Zp(this);
			temp.num = Modulo(num * operand.Value);
			return temp;
		}

		public Zp Calculate(Zp operand, Operation operation)
		{
			switch (operation)
			{
				case Operation.Add:
					return Add(operand);
				case Operation.Mul:
					return Mul(operand);
				case Operation.Sub:
					return Sub(operand);
				case Operation.Div:
					return Div(operand);
				default:
					throw new Exception("Unknown operation: " + operation);
			}
		}

		public int Modulo(int i)
		{
			return Modulo(i, Prime);
		}

		public static int Modulo(int n, int prime)
		{
			n = n % prime;
			return n < 0 ? n + prime : n;
		}

		public static Zp EvalutePolynomialAtPoint(IList<Zp> polynomial, Zp point)
		{
			int evaluation = 0;
			for (int i = 0; i < polynomial.Count; i++)
			{
				evaluation += polynomial[i].Value * NumTheoryUtils.ModPow(point.Value, i, point.Prime);
			}
			return new Zp(point.Prime, evaluation);
		}

		public override string ToString()
		{
			return Convert.ToString(num);
		}

		public override bool Equals(object obj)
		{
			if (!(obj is Zp))
				return false;
			Zp compare = (Zp)obj;
			return Value == compare.Value && Prime == compare.Prime;
		}

		public override int GetHashCode()
		{
			return Value.GetHashCode();
		}

        public bool Equals(Zp other)
        {
            return other != null && Value == other.Value && Prime == other.Prime;
        }

        public int Size
		{
			get { return sizeof(int); }
		}
	}
}